openlibrary/plugins/upstream/utils.py

```diff
@@ -734,6 +734,25 @@ def canonical_url(self):
         """
         readable_path = web.ctx.get('readable_path', web.ctx.path) or ''
         query = web.ctx.query or ''
+
+        if len(query) > 0:
+            queries_to_exclude = ['sort', 'mode', 'v', 'type', 'debug']
+            parameters = query.split('&')
+            pattern = re.compile('.*?(?==)')
```

Comment:
Any time you find yourself (effectively) writing your own parser, you should stop and look for an existing parser. In this case, I'd look towards https://docs.python.org/3/library/urllib.parse.html#module-urllib.parse.

The one tricky nuance will be to code it in such a way that it works for Python 2.7 in addition to Python 3. For that you should look into the `six` package which provides version independent imports/implementations that you can use.

Reply:
I believe `i = web.input(_m='GET')` should get the query params as a dict? But not sure what to do with that from there.

Reply:
@flankstaek this seems like the only blocker. I agree with @tfmorris and @cdrini's comments here.